index.js

```diff
@@ -41,6 +71,143 @@ class GrcSlack extends Base {
+   * Batch log error to slack
+   * @param {string} reqChannel - Slack channel to log the error to, if not provided, the channel from the config will be used
+   * @param {Error} err - Error to log
+   * @param {string} sourceName - Source of the error
+   * @param {Object} payload - Payload to log
+   * @param {...any} extra - Additional information to log
+   */
+  async logErrorEnqueue (reqChannel, err, sourceName, payload, ...extra) {
+    if (!this._errorBatch) {
+      console.error('Error batching not initialized, falling back to direct log')
+      return this.logError(reqChannel, err, sourceName, payload, ...extra)
+    }
+
+    if (!reqChannel) {
+      reqChannel = this.conf.channel
+    }
+
+    try {
+      const errorKey = this._createErrorKey(reqChannel, err, sourceName)
+
+      const now = new Date()
+      let errorEntry = this._errorBatch.get(errorKey)
+
+      if (!errorEntry) {
+        errorEntry = {
+          errorMessage: err.message,
+          sourceName,
+          reqChannel,
+          payloads: [
+            { payload, extras: extra }
+          ],
+          count: 0,
+          firstSeen: now,
+          lastSeen: now
+        }
+        this._errorBatch.set(errorKey, errorEntry)
+      }
+
+      errorEntry.count++
+      errorEntry.lastSeen = now
+      errorEntry.payloads.push({ payload, extras: extra })
+    } catch (e) {
+      console.error('Error batching failed, falling back to direct log', e)
+      await this.logError(reqChannel, err, sourceName, payload, ...extra)
+    }
+  }
+
+  _createErrorKey (reqChannel, err, sourceName) {
+    const errorMsg = err?.message || 'Unknown error'
+    return `${reqChannel}:${sourceName}:${errorMsg}`
+  }
+
+  async _processBatchedErrors () {
+    if (!this._errorBatch || this._errorBatch.cache.length === 0) {
+      return
+    }
+
+    try {
+      const errorsByFunctionNameAndChannel = new Map()
+
+      const allEntries = Object.values(this._errorBatch.cache.cache || {})
+
+      for (const { value: errorEntry } of allEntries) {
+        const groupKey = `${errorEntry.reqChannel}:${errorEntry.sourceName}`
+        if (!errorsByFunctionNameAndChannel.has(groupKey)) {
+          errorsByFunctionNameAndChannel.set(groupKey, [])
+        }
+
+        const errorKey = this._createErrorKey(errorEntry.reqChannel, { message: errorEntry.errorMessage }, errorEntry.sourceName)
+        errorsByFunctionNameAndChannel.get(groupKey).push({ errorKey, ...errorEntry })
+      }
+
+      for (const [groupKey, errors] of errorsByFunctionNameAndChannel) {
+        const [reqChannel, sourceName] = groupKey.split(':')
+        await this._sendBatchedErrorMessage(reqChannel, sourceName, errors)
+      }
+    } catch (e) {
+      console.error('Failed to process batched errors', e)
+    } finally {
+      this._errorBatch.clear()
+    }
+  }
+
+  async _sendBatchedErrorMessage (reqChannel, sourceName, errors) {
+    const totalErrors = errors.reduce((sum, error) => sum + error.count, 0)
```

Comment:
you can keep the counter separately to avoid computing it here, 1 less array pass